chrysom/client.go

```diff
@@ -354,28 +370,56 @@ func (c *Client) Start(ctx context.Context) error {
 		return ErrUndefinedIntervalTicker
 	}
 
+	c.observer.mu.Lock()
+	defer c.observer.mu.Unlock()
+
+	if c.observer.state == running {
+		level.Warn(c.logger).Log(xlog.MessageKey(), "Start called when a listener was still running")
+		return nil
+	}
+
+	c.observer.ticker.Reset(c.observer.pullInterval)
 	go func() {
-		observer := c.observer
-		for range observer.ticker.C {
-			outcome := SuccessOutcome
-			items, err := c.GetItems(observer.bucket, observer.owner)
-			if err == nil {
-				observer.listener.Update(items)
-			} else {
-				outcome = FailureOutcome
-				level.Error(c.logger).Log(xlog.MessageKey(), "Failed to get items for listeners", xlog.ErrorKey(), err)
+		for {
+			select {
+			case <-c.observer.shutdown:
+				return
+			case <-c.observer.ticker.C:
+				outcome := SuccessOutcome
+				items, err := c.GetItems(c.observer.bucket, c.observer.owner)
+				if err == nil {
+					c.observer.listener.Update(items)
+				} else {
+					outcome = FailureOutcome
+					level.Error(c.logger).Log(xlog.MessageKey(), "Failed to get items for listeners", xlog.ErrorKey(), err)
+				}
+				c.observer.pollCount.With(OutcomeLabel, outcome).Add(1)
 			}
-			observer.pollCount.With(OutcomeLabel, outcome).Add(1)
 		}
 	}()
 
+	c.observer.state = running
+
 	return nil
 }
 
+// Stop requests the current listener process to start and waits for its goroutine to complete.
+// Calling Stop() when no process is running is a NoOp.
 func (c *Client) Stop(ctx context.Context) error {
-	if c.observer != nil && c.observer.ticker != nil {
-		c.observer.ticker.Stop()
+	if c.observer == nil || c.observer.ticker == nil {
+		return nil
 	}
+
+	c.observer.mu.Lock()
+	defer c.observer.mu.Unlock()
+
+	if c.observer.state == stopped {
+		return nil
+	}
+
+	c.observer.ticker.Stop()
+	c.observer.shutdown <- struct{}{}
```

Comment:
It's better to actually close a shutdown channel.  When `Start` is invoked, you create the channel.  When `Stop` is invoked, you can `close` the channel and `nil` out the field.

The rationale here is that you don't want to miss state changes.  When a channel is reused for shutdown, it can be possible for the previously started goroutine to miss the change and the currently started goroutine to receive the change.  That's because there is no ordering to how goroutines receive data on channels.